File: ruby/cookbook/recipe_thread.rb
#!/usr/bin/env ruby

module Recipe

    require 'logger'

    class << self

        @@logger = LoggerFactory.getLogger(File.basename(__FILE__))

        def title

            @@logger.debug '# Thread'

        end

        def cook

            secs = [1.0, 2.0, 3.0]

            threads = Array.new

            count = 0
            secs.each do |sec|
                count += 1
                threads << Thread.new(count, sec) do |_count, _sec|
                    Thread.current[:name] = "thread-#{_count}"
                    __sleep(_sec)
                end
            end

            threads.each do |thread|
                thread.join if thread.alive?
            end

            @@logger.debug 'all ended'

        end

        def __sleep(sec)

            @@logger.debug "sleep(#{sec}) started"
            sleep sec
            @@logger.debug "sleep(#{sec}) ended"

        end

    end

end

if __FILE__ == $0

    Recipe.title
    Recipe.cook

end

File: python/cookbook/recipe_thread.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

import threading
import time
from cookbook.util.logging import LoggerFactory

logger = LoggerFactory.getLogger(__name__)

def title():

    logger.debug('# Thread')

def cook():

    secs = [1.0, 2.0, 3.0]

    threads = list()

    count = 0
    for sec in secs:
        count += 1
        threads.append(threading.Thread(target=__sleep, name="thread-%s" % (count), args=(sec,)))

    for thread in threads:
        thread.start()

    for thread in threads:
        if thread.isAlive(): thread.join()

    logger.debug('all ended')

def __sleep(sec):

    logger.debug("sleep(%s) started" % (sec))

    time.sleep(sec)

    logger.debug("sleep(%s) ended" % (sec))

if __name__ == '__main__':

    title()
    cook()
